crates/ev-cli/src/commands/function/deploy.rs

```diff
@@ -0,0 +1,237 @@
+    DeploymentCancelled,
+}
+
+impl CmdOutput for DeployError {
+    fn code(&self) -> String {
+        match self {
+            DeployError::FetchAppFunctions(_) => "function-fetch-functions-error",
+            DeployError::MissingNameField => "function-missing-name-error",
+            DeployError::ValidationError(_) => "function-validation-error",
+            DeployError::FsError(_) => "function-fs-error",
+            DeployError::VersionDeprecated(_, _) => "function-version-deprecated-error",
+            DeployError::VersionWillBeDeprecated(_, _) => {
+                "function-version-will-be-deprecated-error"
+            }
+            DeployError::RecordCreate(_) => "function-record-create-error",
+            DeployError::ZipNotFound => "function-zip-not-found-error",
+            DeployError::FunctionUpload(_) => "function-upload-error",
+            DeployError::DeploymentFailed | DeployError::DeploymentCancelled => {
+                "function-deployment-error"
+            }
+            DeployError::DeploymentStatusFetch(_) => "function-deployment-status-fetch-error",
+        }
+        .to_string()
+    }
+
+    fn exitcode(&self) -> crate::errors::ExitCode {
+        crate::errors::GENERAL
+    }
+}
+
+#[derive(strum_macros::Display, Debug)]
+pub enum DeployMessage {
+    #[strum(to_string = "Function ({uuid}) Deployed Succesfully")]
+    Deployed { uuid: String },
+    #[strum(
+        to_string = "Function deployment initiated successfully. Deployment will continue in the background."
+    )]
+    BackgroundDeployment,
+}
+
+impl CmdOutput for DeployMessage {
+    fn code(&self) -> String {
+        match self {
+            DeployMessage::Deployed { .. } => "function-deployed",
+            DeployMessage::BackgroundDeployment => "function-background-deployment-started",
+        }
+        .to_string()
+    }
+
+    fn exitcode(&self) -> crate::errors::ExitCode {
+        crate::errors::OK
+    }
+}
+
+pub async fn run(args: DeployArgs, auth: BasicAuth) -> Result<DeployMessage, DeployError> {
+    let api_client = papi::EvApiClient::new(auth);
+    let base_args = BaseArgs::parse();
+    let is_quiet = !base_args.verbose || base_args.json;
+
+    crate::fs::validate_function_directory_structure()?;
+    crate::fs::validate_function_toml()?;
+
+    let name = get_current_function_name().map_err(|_| DeployError::MissingNameField)?;
+
+    validators::validate_function_name(&name)?;
+
+    let language = get_current_function_language()?;
+
+    validate_function_language(&language)?;
+
+    let current_date: NaiveDate = Utc::now().date_naive();
+
+    if let Some(&deprecation_date) = LANGUAGE_DEPRECATION_DATE_MAP.get(&language) {
+        return Err(if current_date > deprecation_date {
+            DeployError::VersionDeprecated(deprecation_date, language)
+        } else {
+            DeployError::VersionWillBeDeprecated(deprecation_date, language)
```

Comment:
Is this blocking deployments in advance of deprecation..?

Reply:
yea good catch, just meant to print here